pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go

```diff
@@ -0,0 +1,341 @@
+	return path.Join("projects", spec.Project, "locations", spec.Location)
+}
+
+type Config struct {
+	RootSpec         CertificateAuthoritySpec   `hcl:"root_cert_spec,block"`
+	TrustBundleSpecs []CertificateAuthoritySpec `hcl:"trust_bundle_cert_spec,block"`
+}
+
+type Plugin struct {
+	// gRPC requires embedding either the "Unimplemented" or "Unsafe" stub as
+	// a way of opting in or out of forward build compatibility.
+	upstreamauthority.UnimplementedUpstreamAuthorityServer
+
+	// mu is a mutex that protects the configuration. Plugins may at some point
+	// need to support hot-reloading of configuration (by receiving another
+	// call to Configure). So we need to prevent the configuration from
+	// being used concurrently and make sure it is updated atomically.
+	mu sync.Mutex
+	c  *Config
+
+	log hclog.Logger
+}
+
+// These are compile time assertions that the plugin matches the interfaces the
+// catalog requires to provide the plugin with a logger and host service
+// broker as well as the UpstreamAuthority itself.
+var _ catalog.NeedsLogger = (*Plugin)(nil)
+var _ upstreamauthority.UpstreamAuthorityServer = (*Plugin)(nil)
+
+func New() *Plugin {
+	return &Plugin{}
+}
+
+// SetLogger will be called by the catalog system to provide the plugin with
+// a logger when it is loaded. The logger is wired up to the SPIRE core
+// logger
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Mints an X.509 CA and responds with the signed X.509 CA certificate
+// chain and upstream X.509 roots. If supported by the implementation,
+// subsequent responses on the stream contain upstream X.509 root updates,
+// otherwise the RPC is completed after sending the initial response.
+//
+// Implementation note:
+// The stream should be kept open in the face of transient errors
+// encountered while tracking changes to the upstream X.509 roots as SPIRE
+// core will not reopen a closed stream until the next X.509 CA rotation.
+func (p *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
+	ctx := stream.Context()
+
+	minted, err := p.mintX509CA(ctx, request.Csr, request.PreferredTtl)
+	if err != nil {
+		return err
+	}
+
+	return stream.Send(minted)
+}
+
+func (p *Plugin) Configure(ctx context.Context, req *plugin.ConfigureRequest) (*plugin.ConfigureResponse, error) {
+	// Parse HCL config payload into config struct
+	config := new(Config)
+	if err := hcl.Decode(config, req.Configuration); err != nil {
+		return nil, pluginErr.New("unable to decode configuration: %v", err)
+	}
+
+	// Swap out the current configuration with the new configuration
+	p.setConfig(config)
+
+	return &plugin.ConfigureResponse{}, nil
+}
+
+func (p *Plugin) GetPluginInfo(ctx context.Context, req *plugin.GetPluginInfoRequest) (*plugin.GetPluginInfoResponse, error) {
+	return &plugin.GetPluginInfoResponse{}, nil
+}
+
+func (p *Plugin) getConfig() (*Config, error) {
```

Comment:
do we use `getConfig()` anywhere?

Reply:
It should be used instead of reaching into `p.c` directly. I've commented where that is taking place. Ideally this is called at the top of `MintX509CA` and then the returned configuration passed into subsequent helpers that satisfy the RPC.